src/dict.c

```diff
@@ -509,6 +544,8 @@ dictEntry *dictInsertAtPosition(dict *d, void *key, void *position) {
             /* Allocate an entry without value. */
             entry = createEntryNoValue(key, *bucket);
         }
+    } else if (d->type->embedded_entry) {
+        entry = createEmbeddedEntry(key, *bucket, d->type);
```

Comment:
The way `dictEntry` is used has no type safety. Admittedly, this is not a new issue but the addition of `embeddedDictEntry` is making it worse. The following code path looks problematic to me. can you please double check?

`setGenericComamd() -> setKey() -> dbAdd() -> dbAddInsternal() -> kvstoreDictSetKey() -> dictSetKey()`

I don't see`dictSetKey` getting patched to handle this new embedded dict entry.

I would suggest making `dictEntry` an opaque struct next and force every function go through an inline accessor function/macro. I think this is the only certain way to ensure we don't accidentally use the wrong data type.

Reply:
not in dict.c.

Reply:
I guess I don't agree it should be opaque in dict.c.